File: tools/oracle/src/geometry.rs
//! The camera, the image rectangle, and HLD 25.1's geometry bound.
//!
//! **The canvas scale is read, never re-derived.** `canvasScale` in
//! `tools/oracle/src/params.mjs` computes canvas pixels per source pixel from
//! the camera's own `parallelScale` and the row and column spacing, and F-X007
//! publishes it on every stack sidecar as `canvasPixelsPerSourcePixel`. That
//! was decision 13 of F-011's design round, taken so this file would not hold
//! a second copy of that derivation. HLD section 18's "do not let a second
//! copy of this logic appear anywhere" is about the LUT chain and the
//! reasoning generalises.
//!
//! What this file derives is the RECTANGLE that scale implies, which is the
//! line between the picture and the letterbox and the denominator of
//! `informativeFraction`. It is not the region the bias bullet is averaged
//! over: 25.1 names the informative region, which is the subset of this
//! rectangle that is not clipped to the same display extreme on both sides.
//!
//! **There is no float to integer cast here.** `as` on a float truncates
//! toward zero and saturates silently out of range, and all three cast lints
//! are denied. The rectangle's edges are found by comparing pixel centres
//! against the extent, which is a scan over at most a canvas width of
//! candidates and is exact.

use crate::frame::{FrameError, Rect};
use crate::tolerance::{CANVAS_TOLERANCE_PIXELS, WORLD_TOLERANCE_MM};

/// One field that is outside a bound, with both readings and the bound it
/// broke, so the report can name it rather than saying "geometry differs".
#[derive(Clone, Debug)]
pub struct Divergence {
    pub field: String,
    pub reference: f64,
    pub candidate: f64,
    pub difference: f64,
    pub bound: f64,
}

/// The camera block of a sidecar. `view_plane_normal` is present on a volume
/// reformat and absent on a stack frame, and a reformat is meaningless without
/// it.
#[derive(Clone, Debug)]
pub struct Camera {
    pub position: [f64; 3],
    pub focal_point: [f64; 3],
    pub view_up: [f64; 3],
    pub parallel_scale: f64,
    pub view_plane_normal: Option<[f64; 3]>,
}

fn compare(field: &str, a: f64, b: f64, bound: f64, into: &mut Vec<Divergence>) {
    let difference = (b - a).abs();
    // A NaN comparison is false in both directions, so a NaN would slip past a
    // `>` test silently. It is reported as a divergence with an infinite
    // difference instead, because a camera that carries one is not a camera.
    let outside = difference.is_nan() || difference > bound;
    if outside {
        into.push(Divergence {
            field: field.to_owned(),
            reference: a,
            candidate: b,
            difference,
            bound,
        });
    }
}

fn compare_triple(field: &str, a: [f64; 3], b: [f64; 3], bound: f64, into: &mut Vec<Divergence>) {
    for (index, (left, right)) in a.iter().zip(b.iter()).enumerate() {
        compare(&format!("{field}[{index}]"), *left, *right, bound, into);
    }
}

/// 25.1's "world coordinates within 1e-6 mm", applied to the camera.
///
/// `view_up` and `view_plane_normal` are direction cosines rather than
/// millimetres, and they are held to the same number deliberately: a unit
/// vector wrong by more than a millionth is a different orientation, and 25.1
/// states no second bound for a direction.
#[must_use]
pub fn world_divergences(reference: &Camera, candidate: &Camera) -> Vec<Divergence> {
    let mut found = Vec::new();
    compare_triple(
        "camera.position",
        reference.position,
        candidate.position,
        WORLD_TOLERANCE_MM,
        &mut found,
    );
    compare_triple(
        "camera.focalPoint",
        reference.focal_point,
        candidate.focal_point,
        WORLD_TOLERANCE_MM,
        &mut found,
    );
    compare_triple(
        "camera.viewUp",
        reference.view_up,
        candidate.view_up,
        WORLD_TOLERANCE_MM,
        &mut found,
    );
    compare(
        "camera.parallelScale",
        reference.parallel_scale,
        candidate.parallel_scale,
        WORLD_TOLERANCE_MM,
        &mut found,
    );
    match (reference.view_plane_normal, candidate.view_plane_normal) {
        (Some(a), Some(b)) => compare_triple(
            "camera.viewPlaneNormal",
            a,
            b,
            WORLD_TOLERANCE_MM,
            &mut found,
        ),
        (None, None) => {}
        (a, b) => found.push(Divergence {
            field: "camera.viewPlaneNormal".to_owned(),
            reference: if a.is_some() { 1.0 } else { 0.0 },
            candidate: if b.is_some() { 1.0 } else { 0.0 },
            difference: f64::INFINITY,
            bound: WORLD_TOLERANCE_MM,
        }),
    }
    found
}

/// The image's extent on the canvas, in canvas pixels, before it is turned
/// into whole pixels.
#[derive(Clone, Debug)]
pub struct CanvasExtent {
    pub x0: f64,
    pub y0: f64,
    pub width: f64,
    pub height: f64,
}

impl CanvasExtent {
    /// The extent of an image of the given canvas size, centred in the canvas.
    ///
    /// `resetCamera` fits the image by its extent in millimetres and centres
    /// it, which is what makes the offset half the leftover in each direction.
    #[must_use]
    pub fn centred(width: f64, height: f64, canvas_width: u32, canvas_height: u32) -> Self {
        Self {
            x0: (f64::from(canvas_width) - width) / 2.0,
            y0: (f64::from(canvas_height) - height) / 2.0,
            width,
            height,
        }
    }

    /// The whole canvas pixels this extent covers.
    ///
    /// A pixel belongs to the image when its CENTRE lies inside the extent, so
    /// pixel `i` is inside when `i + 0.5 >= x0` and `i + 0.5 < x0 + width`.
    /// The alternatives were tried and both are wrong in a way that is
    /// invisible: flooring the offset admits a letterbox column, and ceiling
    /// the far edge admits one on the other side. At 512 rows either is 512
    /// background pixels counted as picture.
    ///
    /// **The harm is not to the bias bullet.** Both sides paint the declared
    /// clear colour in the letterbox, so a wrongly admitted column is black on
    /// both sides, which is clipped to the same extreme and therefore
    /// uninformative, and it never reaches the bias denominator at all. The
    /// two things it does break:
    ///
    /// 1. `informativeFraction` is informative pixels over image-rectangle
    ///    pixels, so an admitted column inflates the denominator alone and
    ///    pushes the fraction down towards `INFORMATIVE_FRACTION_FLOOR`. The
    ///    lowest fraction among the views the identity run does not call weak
    ///    is 0.10074 against a floor of 0.10, so the margin is 0.0007 and 512
    ///    of 262144 pixels is 0.00195 of the rectangle. Reproduce the fraction
    ///    from `informativeFraction` in the `compare.json` that
    ///    `./target/release/ocelli-compare identity` writes.
    /// 2. The `letterbox-only` qualifier fires only when the image region
    ///    carries NO difference and the background carries one. A fit error in
    ///    a column that should have been letterbox then lands inside the image
    ///    region, the qualifier cannot fire, and a difference in the fit is
    ///    attributed to the picture and to us.
    ///
    /// The result is clamped to the canvas, because the published scale
    /// carries floating point noise and a frame fitted DOWN into the canvas
    /// can put its far edge a millionth of a pixel outside.
    #[must_use]
    pub fn rect(&self, canvas_width: u32, canvas_height: u32) -> Rect {
        let x_start = first_centre_at_or_after(self.x0, canvas_width);
        let x_end = first_centre_at_or_after(self.x0 + self.width, canvas_width);
        let y_start = first_centre_at_or_after(self.y0, canvas_height);
        let y_end = first_centre_at_or_after(self.y0 + self.height, canvas_height);
        Rect {
            x0: x_start,
            y0: y_start,
            width: x_end.saturating_sub(x_start),
            height: y_end.saturating_sub(y_start),
        }
    }
}

/// The first pixel index in `0..limit` whose centre is at or after `edge`, or
/// `limit` when there is none.
///
/// A scan rather than a cast. `limit` is a canvas dimension, 512 today, so the
/// cost is nothing and the arithmetic is exact for every input including a
/// negative or a non-finite edge.
fn first_centre_at_or_after(edge: f64, limit: u32) -> u32 {
    if edge.is_nan() {
        return limit;
    }
    (0..limit)
        .find(|index| f64::from(*index) + 0.5 >= edge)
        .unwrap_or(limit)
}

/// The stack case: the image's canvas extent from the published scale.
///
/// # Errors
/// Never today. The signature carries a `Result` because the caller's other
/// extent sources can fail and a uniform shape keeps the call sites readable.
pub fn stack_extent(
    columns: u32,
    rows: u32,
    horizontal: f64,
    vertical: f64,
    canvas_width: u32,
    canvas_height: u32,
) -> Result<CanvasExtent, FrameError> {
    Ok(CanvasExtent::centred(
        f64::from(columns) * horizontal,
        f64::from(rows) * vertical,
        canvas_width,
        canvas_height,
    ))
}

/// 25.1's "canvas coordinates within a quarter pixel", applied to the two
/// sides' image extents.
///
/// The offsets are compared as well as the sizes, because two extents of the
/// same size in different places are two different pictures and a size-only
/// comparison would pass a translated frame.
#[must_use]
pub fn canvas_divergences(reference: &CanvasExtent, candidate: &CanvasExtent) -> Vec<Divergence> {
    let mut found = Vec::new();
    compare(
        "imageExtent.x0",
        reference.x0,
        candidate.x0,
        CANVAS_TOLERANCE_PIXELS,
        &mut found,
    );
    compare(
        "imageExtent.y0",
        reference.y0,
        candidate.y0,
        CANVAS_TOLERANCE_PIXELS,
        &mut found,
    );
    compare(
        "imageExtent.width",
        reference.width,
        candidate.width,
        CANVAS_TOLERANCE_PIXELS,
        &mut found,
    );
    compare(
        "imageExtent.height",
        reference.height,
        candidate.height,
        CANVAS_TOLERANCE_PIXELS,
        &mut found,
    );
    found
}

/// The reformat case: 25.1's canvas bullet applied to a frame with no source
/// pixel grid.
///
/// A reformat plane is a cut through a volume, so there is no magnification
/// factor and no image rectangle inside the canvas. What the two sides must
/// agree on is the SCALE, `reformat.millimetresPerCanvasPixel`, and the
/// quarter-pixel bound is applied where it bites hardest: at the canvas edge,
/// half a canvas away from the focal point, where a scale disagreement has
/// accumulated the most.
#[must_use]
pub fn reformat_scale_divergences(
    reference_mm_per_pixel: f64,
    candidate_mm_per_pixel: f64,
    canvas_height: u32,
) -> Vec<Divergence> {
    let half = f64::from(canvas_height) / 2.0;
    let drift_mm = (candidate_mm_per_pixel - reference_mm_per_pixel).abs() * half;
    let drift_pixels = if reference_mm_per_pixel > 0.0 {
        drift_mm / reference_mm_per_pixel
    } else {
        f64::INFINITY
    };
    let mut found = Vec::new();
    if drift_pixels.is_nan() || drift_pixels > CANVAS_TOLERANCE_PIXELS {
        found.push(Divergence {
            field: "reformat.millimetresPerCanvasPixel".to_owned(),
            reference: reference_mm_per_pixel,
            candidate: candidate_mm_per_pixel,
            difference: drift_pixels,
            bound: CANVAS_TOLERANCE_PIXELS,
        });
    }
    found
}

#[cfg(test)]
mod tests {
    use super::{Camera, CanvasExtent, reformat_scale_divergences, world_divergences};

    /// A camera carrying a `NaN` is not a camera, and a `>` test against a
    /// bound is false in both directions for one. It is reported rather than
    /// passed.
    #[test]
    fn a_nan_in_the_camera_is_a_divergence_and_not_a_pass() {
        let base = Camera {
            position: [0.0, 0.0, 0.0],
            focal_point: [0.0, 0.0, 0.0],
            view_up: [0.0, -1.0, 0.0],
            parallel_scale: 16.0,
            view_plane_normal: None,
        };
        let mut broken = base.clone();
        broken.parallel_scale = f64::NAN;
        assert_eq!(world_divergences(&base, &broken).len(), 1);
    }

    /// A stack sidecar carries no `viewPlaneNormal` and a reformat one does.
    /// Comparing a frame that has one against a frame that does not is a
    /// divergence rather than a field quietly skipped.
    #[test]
    fn a_missing_view_plane_normal_on_one_side_only_is_a_divergence() {
        let base = Camera {
            position: [0.0, 0.0, 0.0],
            focal_point: [0.0, 0.0, 0.0],
            view_up: [0.0, -1.0, 0.0],
            parallel_scale: 16.0,
            view_plane_normal: None,
        };
        let mut reformat = base.clone();
        reformat.view_plane_normal = Some([0.0, 0.0, -1.0]);
        assert_eq!(world_divergences(&base, &reformat).len(), 1);
        assert!(world_divergences(&reformat, &reformat).is_empty());
    }

    /// The reformat scale bound, at the AXIAL MR frame's own numbers.
    /// 0.703125 mm per canvas pixel over a 512 pixel canvas: a quarter pixel
    /// at the canvas edge is a scale change of
    /// 0.25 * 0.703125 / 256 = 0.000687 mm per pixel.
    #[test]
    fn the_reformat_scale_bound_is_a_quarter_pixel_at_the_canvas_edge() {
        let base = 0.703_125_f64;
        assert!(reformat_scale_divergences(base, base + 0.000_6, 512).is_empty());
        assert_eq!(
            reformat_scale_divergences(base, base + 0.000_8, 512).len(),
            1
        );
    }

    /// An extent that lies entirely outside the canvas produces an empty
    /// rectangle rather than a rectangle that wraps.
    #[test]
    fn an_extent_off_the_canvas_produces_an_empty_rectangle() {
        let extent = CanvasExtent {
            x0: 600.0,
            y0: 0.0,
            width: 10.0,
            height: 512.0,
        };
        let rect = extent.rect(512, 512);
        assert!(rect.is_empty());
    }
}

File: tools/oracle/src/geometry.mjs
// The harness's own reading of a series' geometry, from the files.
//
// PS3.3 C.7.6.2.1.1 is the normative source for every line below. It gives
// Image Orientation (Patient) (0020,0037) as the direction cosines of the
// first ROW and the first COLUMN of the image, and Pixel Spacing (0028,0030)
// as `[between rows, between columns]`. Image Position (Patient) (0020,0032)
// is the position of the centre of the first transmitted voxel.
//
// **This module never reads a cornerstone3D module.** It is the second opinion
// the sidecar already carries for pixel metadata, applied to geometry, and it
// exists for the same reason: a harness that transcribed the reference's own
// answer could not show the reference getting it wrong, which is exactly the
// case on a non-uniform series. See docs/lld/oracle.md.
//
// Two derivations in here are the ones a reviewer should check against the
// standard rather than against the comment above them, because both have a
// plausible wrong form:
//
//  1. **Through-plane spacing is the difference of PROJECTED
//     `ImagePositionPatient` values**, projected onto the cross product of the
//     two direction cosines. Never `SpacingBetweenSlices` (0018,0088), which is
//     Type 3 and frequently wrong, and never `SliceThickness` (0018,0050),
//     which is Type 2 and describes the slab rather than the step.
//     `scripts/corpus_synth.py`'s `case_series` omits (0018,0088) deliberately
//     and gives both series a (0018,0050) of 2.5, so a reader that took either
//     tag would answer 2.5 for a series whose gaps are 2.5, 3.75 and 1.25.
//  2. **`PixelSpacing[0]` multiplies the COLUMN direction cosine.** Advancing
//     one row moves along the column direction, which is the second triplet of
//     (0020,0037). The corpus's deliberately non-square `[0.5, 0.25]` is what
//     makes the transposition visible, and `tests/geometry_test.mjs` asserts
//     it by hand.
//
// HLD section 19 declares `spacing: [f64; 3]` in millimetres and `origin` as
// "IPP of the first slice", and says NOTHING about how the through-plane
// element is derived. That gap is real and is recorded in docs/lld/oracle.md
// so the story that builds the volume can raise it with evidence. This module
// cites PS3.3 directly instead.
//
// Every number here is `World` in HLD section 16's sense: DICOM patient
// coordinates, LPS, millimetres.

/** HLD 25.1: "Geometry: world coordinates within 1e-6 mm". */
export const GEOMETRY_TOLERANCE_MM = 1e-6;

/** The cross product, in the order `row x col`. */
export function cross(a, b) {
  return [
    a[1] * b[2] - a[2] * b[1],
    a[2] * b[0] - a[0] * b[2],
    a[0] * b[1] - a[1] * b[0],
  ];
}

/** The dot product of two three-vectors. */
export function dot(a, b) {
  return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
}

/** A three-vector scaled by a number. */
export function scale(v, k) {
  return [v[0] * k, v[1] * k, v[2] * k];
}

/** The Euclidean length of a three-vector. */
export function norm(v) {
  return Math.sqrt(dot(v, v));
}

/** Whether two numbers agree within HLD 25.1's geometry tolerance. */
export function within(a, b, toleranceMm = GEOMETRY_TOLERANCE_MM) {
  return Math.abs(a - b) <= toleranceMm;
}

/** Whether two vectors agree element-wise within the geometry tolerance. */
export function vectorsWithin(a, b, toleranceMm = GEOMETRY_TOLERANCE_MM) {
  if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
    return false;
  }
  return a.every((value, index) => within(value, b[index], toleranceMm));
}

function numbers(value, count, what, path) {
  if (
    !Array.isArray(value) ||
    value.length !== count ||
    value.some((entry) => typeof entry !== "number" || !Number.isFinite(entry))
  ) {
    throw new Error(
      `${path}: ${what} is ${JSON.stringify(value)} and PS3.3 C.7.6.2.1.1 ` +
        `requires ${count} finite decimal values. A series member with no ` +
        `geometry cannot be placed, and placing it anyway is how a volume ` +
        `comes out plausible and wrong.`,
    );
  }
  return value;
}

/**
 * The median of a list of numbers.
 *
 * An even count takes the mean of the two middle values, which is the usual
 * definition and is stated here because a nine-gap series takes the odd branch
 * and the even one would otherwise be written by whoever first met it.
 */
export function median(values) {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 1
    ? sorted[middle]
    : (sorted[middle - 1] + sorted[middle]) / 2;
}

/**
 * Measure one series' geometry from its members' own attributes.
 *
 * @param {Array<{path: string, imagePositionPatient: number[],
 *   imageOrientationPatient: number[], pixelSpacing: number[]}>} members
 * @param {{toleranceMm?: number}} options
 */
export function measureSubject(members, options = {}) {
  const toleranceMm = options.toleranceMm ?? GEOMETRY_TOLERANCE_MM;
  if (!Array.isArray(members) || members.length < 2) {
    throw new Error(
      `a series needs at least two members to have a gap, and this one has ` +
        `${Array.isArray(members) ? members.length : 0}`,
    );
  }

  const first = members[0];
  const orientation = numbers(
    first.imageOrientationPatient,
    6,
    "ImageOrientationPatient (0020,0037)",
    first.path,
  );
  const spacing = numbers(
    first.pixelSpacing,
    2,
    "PixelSpacing (0028,0030)",
    first.path,
  );

  const rowCosine = orientation.slice(0, 3);
  const columnCosine = orientation.slice(3, 6);
  const normal = cross(rowCosine, columnCosine);

  // PS3.3 C.7.6.2.1.1 gives (0020,0037) as direction COSINES, so both triplets
  // are unit vectors and the standard requires them to be orthogonal. Their
  // cross product is then a unit vector too, and the projection below is a
  // distance in millimetres only because it is. A normal of another length
  // would scale every gap by an amount nobody declared, so this is refused
  // rather than normalised: a file whose cosines are not cosines is a file
  // whose geometry is not trustworthy, and silently fixing it is the quiet
  // kind of wrong.
  if (!within(norm(rowCosine), 1, toleranceMm) ||
      !within(norm(columnCosine), 1, toleranceMm) ||
      !within(norm(normal), 1, toleranceMm)) {
    throw new Error(
      `${first.path}: ImageOrientationPatient ` +
        `${JSON.stringify(orientation)} is not two perpendicular vectors of ` +
        `unit length (|row| ${norm(rowCosine)}, |col| ${norm(columnCosine)}, ` +
        `|row x col| ${norm(normal)}). PS3.3 C.7.6.2.1.1 gives them as ` +
        `direction cosines, and the projected distance below is in ` +
        `millimetres only because they are.`,
    );
  }

  // Every member must agree, because one volume carries one orientation and
  // one in-plane spacing. A series that does not agree is not one volume, and
  // averaging the disagreement away is the same defect as averaging the gaps.
  const placed = members.map((entry) => {
    const memberOrientation = numbers(
      entry.imageOrientationPatient,
      6,
      "ImageOrientationPatient (0020,0037)",
      entry.path,
    );
    const memberSpacing = numbers(
      entry.pixelSpacing,
      2,
      "PixelSpacing (0028,0030)",
      entry.path,
    );
    const position = numbers(
      entry.imagePositionPatient,
      3,
      "ImagePositionPatient (0020,0032)",
      entry.path,
    );
    if (!vectorsWithin(memberOrientation, orientation, toleranceMm)) {
      throw new Error(
        `${entry.path}: ImageOrientationPatient is ` +
          `${JSON.stringify(memberOrientation)} and ${first.path} declares ` +
          `${JSON.stringify(orientation)}. One volume carries one ` +
          `orientation, so a series that disagrees is not one volume.`,
      );
    }
    if (!vectorsWithin(memberSpacing, spacing, toleranceMm)) {
      throw new Error(
        `${entry.path}: PixelSpacing is ${JSON.stringify(memberSpacing)} and ` +
          `${first.path} declares ${JSON.stringify(spacing)}. One volume ` +
          `carries one in-plane spacing.`,
      );
    }
    return { path: entry.path, position, projectionMm: dot(position, normal) };
  });

  // Ordered by the projected position and by nothing else. Not by
  // InstanceNumber, which is Type 2 and is a label rather than a geometry, and
  // not by file name.
  const order = [...placed].sort((a, b) => a.projectionMm - b.projectionMm);
  const projectionsMm = order.map((entry) => entry.projectionMm);

  const gapsMm = [];
  for (let index = 1; index < projectionsMm.length; index += 1) {
    const gap = projectionsMm[index] - projectionsMm[index - 1];
    if (gap <= toleranceMm) {
      // The two paths and NOT their positions. This message is the one refusal
      // here that a REAL corpus row reaches, and `corpus_check.py`'s convention
      // is that a real row reports the relative path and the attribute name
      // and never the value. The paths identify which two members collided,
      // which is the whole diagnostic, and the coordinate adds nothing that a
      // sidecar does not already carry into ignored output.
      throw new Error(
        `${order[index - 1].path} and ${order[index].path} project to the ` +
          `same position on the slice normal, within ${toleranceMm} mm. Two ` +
          `members at one position is not a gap of zero, it is a series that ` +
          `cannot be a volume.`,
      );
    }
    gapsMm.push(gap);
  }

  const meanGapMm =
    gapsMm.reduce((total, gap) => total + gap, 0) / gapsMm.length;
  const maxDeviationFromMeanMm = Math.max(
    ...gapsMm.map((gap) => Math.abs(gap - meanGapMm)),
  );

  return {
    order,
    normal,
    rowCosine,
    columnCosine,
    pixelSpacing: spacing,
    projectionsMm,
    gapsMm,
    meanGapMm,
    medianGapMm: median(gapsMm),
    minGapMm: Math.min(...gapsMm),
    maxGapMm: Math.max(...gapsMm),
    maxDeviationFromMeanMm,
    voxelAxes: {
      // PS3.3 C.7.6.2.1.1: PixelSpacing is [between rows, between columns].
      // Advancing one COLUMN moves along the ROW direction cosine by the
      // between-columns spacing, and advancing one ROW moves along the COLUMN
      // direction cosine by the between-rows spacing. Getting these two the
      // other way round is the transposition the corpus's non-square
      // [0.5, 0.25] exists to catch.
      columnStepMm: scale(rowCosine, spacing[1]),
      rowStepMm: scale(columnCosine, spacing[0]),
      // The MEAN gap, which is what an averaging builder uses. Recorded under
      // a name that says it is a step so a reader can see that on a
      // non-uniform series no interior slice sits where it says.
      sliceStepMm: scale(normal, meanGapMm),
    },
  };
}
